[run-webkit-tests] Route media tests to a mutually exclusive group

JonWBedard opened the pull request:
#### 7b0ca1365193adda601a2af7d1c2a53a913b21c6
<pre>
[run-webkit-tests] Route media tests to a mutually exclusive group
<a href="https://bugs.webkit.org/show_bug.cgi?id=273797">https://bugs.webkit.org/show_bug.cgi?id=273797</a>
<a href="https://rdar.apple.com/127629496">rdar://127629496</a>

Reviewed by Sam Sneddon.

We have evidence that media tests can interfere with one another. We should run media
tests in a dedicated worker so that these tests never interfere with other media tests.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
(LayoutTestRunner.run_tests): Use groups as defined by the current port. Dispatch grouped shards
first, so that we aren&apos;t waiting on groups at the end of the test run.
(TestShard.__init__): Support empty test list.
* Tools/Scripts/webkitpy/port/base.py:
(Port.sharding_groups): Return all groups and their callback definitions.
(Port.group_for_shard): Given a shard, return the group that shard is part of.
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.sharding_groups): Add dedicated &apos;media&apos; and &apos;mediacapture&apos; groups for all shards
containing a media or mediacapture tests.
* Tools/Scripts/webkitpy/port/darwin_testcase.py:
(DarwinTest.test_lanes):
* Tools/Scripts/webkitpy/port/port_testcase.py:
(PortTestCase.test_lanes):

Canonical link: <a href="https://commits.webkit.org/279079@main">https://commits.webkit.org/279079@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db2d0514a54f257385573bad8be10372f4ae5a64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55555 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3004 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54586 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2702 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42532 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1921 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45122 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23607 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/52125 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26483 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2396 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1163 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48388 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57151 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27407 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2588 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49924 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28640 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45241 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49164 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11450 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29552 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28385 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -512,6 +514,14 @@ def __init__(self, name, test_inputs):
+        for i in self.test_inputs:
+            if s in i.test.test_path:
+                return True
+        return False
```
This feels… kinda surprising, honestly? I wouldn't expect `"a" in shard` to be true for a shard containing `TestInput(test_path="foo/testable.html")`?

JonWBedard replied:
What is the suggestion here, then? We want, for example, `imported/w3c/web-platform-tests/media/` to be run in the `media` shard, right?

gsnedders commented on Tools/Scripts/webkitpy/port/darwin.py:
```diff
@@ -53,6 +53,11 @@ def __init__(self, host, port_name, **kwargs):
+    def sharding_groups(self):
+        return dict(
+            media=lambda shard: 'media/' in shard or 'webaudio/' in shard,
+        )
```
```suggestion
        return {
            "media": lambda shard: 'media/' in shard or 'webaudio/' in shard,
        }
```

(We shouldn't be using the dict constructor where we could use a literal)

gsnedders commented on Tools/Scripts/webkitpy/port/base.py:
```diff
@@ -261,6 +261,15 @@ def all_baseline_search_paths(self, device_type=None):
+        return dict()
+
+    def group_for_shard(self, shard):
+        for group, fltr in self.sharding_groups().items():
```
`fltr` is… a shortened form of `filter`? Are we doing this just to avoid overriding the built-in `filter`? I'd rather something like `filter_fn`, rather than making me have to think.

gsnedders commented on Tools/Scripts/webkitpy/port/base.py:
```diff
@@ -261,6 +261,15 @@ def all_baseline_search_paths(self, device_type=None):
         return paths
 
+    def sharding_groups(self):
+        return dict()
```
```suggestion
        return {}
```

gsnedders commented on Tools/Scripts/webkitpy/port/darwin_testcase.py:
```diff
@@ -40,6 +40,13 @@ def assert_skipped_file_search_paths(self, port_name, expected_paths, use_webkit
         port = self.make_port(port_name=port_name, options=MockOptions(webkit_test_runner=use_webkit2))
         self.assertEqual(port._skipped_file_search_paths(), expected_paths)
 
+    def test_lanes(self):
```
```suggestion
    def test_sharding_groups(self):
```

gsnedders commented on Tools/Scripts/webkitpy/port/port_testcase.py:
```diff
@@ -127,6 +127,13 @@ def make_port(self, host=None, port_name=None, options=None, os_name=None, os_ve
         port._config.build_directory = lambda configuration, for_host=False: '/mock-build'
         return port
 
+    def test_lanes(self):
```
```suggestion
    def test_sharding_groups(self):
```

gsnedders commented on Tools/Scripts/webkitpy/port/darwin.py:
```diff
@@ -53,6 +53,11 @@ def __init__(self, host, port_name, **kwargs):
 
+    def sharding_groups(self):
+        return dict(
+            media=lambda shard: 'media/' in shard or 'webaudio/' in shard,
```
but not, e.g., `media-source` or `media-playback-quality` or `media-elements`?

JonWBedard replied:
Yes.

The issue here is that if we include _all_ media tests, we end up extending test time. Put another way, we actually have too many media tests to run all of them serially. I'm going to try tweaking this logic a bit, though. We might be able to get away with a few different media buckets and still greatly improve stability because we're limiting the maximum number of media tests running concurrently.

gsnedders commented on Tools/Scripts/webkitpy/port/darwin_testcase.py:
```diff
@@ -40,6 +40,13 @@ def assert_skipped_file_search_paths(self, port_name, expected_paths, use_webkit
+        self.assertEqual(sorted(port.sharding_groups().keys()), ['media'])
+        self.assertEqual('media', port.group_for_shard('media/something'))
+        self.assertEqual('media', port.group_for_shard('webaudio/something'))
+        self.assertEqual(None, port.group_for_shard('fast/something'))
```
can we also test:

* `fast/media/something`,
* the WPT case,
* `media-session/something`?

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -512,6 +531,9 @@ def __init__(self, name, test_inputs):
         self.test_inputs = test_inputs
         self.needs_servers = test_inputs[0].needs_servers
 
+    def __contains__(self, s):
```
I still feel like this is a surprising implementation of `__contains__`.

In my mind, it makes sense for `TestShard` to be a [`Collection`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Collection), with an iterator like `def __iter__(self): return iter(self.test_inputs)`, and note the default implementation of [`__contains__`](https://docs.python.org/3/reference/datamodel.html#object.__contains__) which just checks if `item` is in the iterator.

As such, I'd expect something like `def __contains__(self, s: TestInput)`, rather than `def __contains__(self, s: str)`. (This is technically incorrect, because `s` always has a type `object` in the parent.)

gsnedders commented on Tools/Scripts/webkitpy/port/darwin_testcase.py:
```diff
@@ -40,6 +40,16 @@ def assert_skipped_file_search_paths(self, port_name, expected_paths, use_webkit
+        self.assertEqual('media', port.group_for_shard('fast/media/something'))
+        self.assertEqual('media', port.group_for_shard('media-session/something'))
+        self.assertEqual('media', port.group_for_shard('imported/mediacapture-fromelement/something'))
+        self.assertEqual(None, port.group_for_shard('fast/something'))
```
Also `something/media` is probably worth testing here, so we know that behaviour there was deliberate.

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -510,7 +529,7 @@ class TestShard(object):
         self.name = name
         self.test_inputs = test_inputs
-        self.needs_servers = test_inputs[0].needs_servers
+        self.needs_servers = test_inputs[0].needs_servers if len(test_inputs) > 0 else False
```
When are we ever creating a TestShard with zero inputs? Shouldn't this be an error?

JonWBedard replied:
We create `TestShard` object with zero inputs in testing. And it's much easier to handle this here than to fully define `test_input` objects to pass into our `TestShard`s.

gsnedders reviewed (approved):
This looks fine, except I do question whether we should (now) allow zero-length shards

JonWBedard committed:
[run-webkit-tests] Route media tests to a mutually exclusive group
https://bugs.webkit.org/show_bug.cgi?id=273797
rdar://127629496

Reviewed by Sam Sneddon.

We have evidence that media tests can interfere with one another. We should run media
tests in a dedicated worker so that these tests never interfere with other media tests.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
(LayoutTestRunner.run_tests): Use groups as defined by the current port. Dispatch grouped shards
first, so that we aren't waiting on groups at the end of the test run.
(TestShard.__init__): Support empty test list.
* Tools/Scripts/webkitpy/port/base.py:
(Port.sharding_groups): Return all groups and their callback definitions.
(Port.group_for_shard): Given a shard, return the group that shard is part of.
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.sharding_groups): Add dedicated 'media' and 'mediacapture' groups for all shards
containing a media or mediacapture tests.
* Tools/Scripts/webkitpy/port/darwin_testcase.py:
(DarwinTest.test_lanes):
* Tools/Scripts/webkitpy/port/port_testcase.py:
(PortTestCase.test_lanes):

Canonical link: https://commits.webkit.org/279079@main
